src/Combinatorics/Matroids/realization_space.jl

```diff
@@ -0,0 +1,989 @@
+        phi=hom(R, R, a->a, gens(R))
+    end
+
+    output = small_reduce_one_step(Igens, phi, elim, fullyReduced)
+    output isa String && return "Not Realizable 0 in Semigroup"
+    
+    (Igens, phi, elim, fullyReduced) = output    
+    !fullyReduced && return small_reduce_full_rec(Igens, phi, elim, fullyReduced)
+    
+    R = domain(phi)
+    x = gens(R)    
+    cR = coefficient_ring(R)
+    
+    if length(elim) == length(x)
+        S = cR
+    else
+        S, y = polynomial_ring(cR, length(x) - length(elim))
+    end        
+    z=[]
+    j=1
+    for ele in x
+        if ele in elim
+            push!(z,0)
+        else
+            push!(z, y[j])
+            j+=1
+        end
+    end
+    phi2 = hom(R, S, a->a, z)
+    phif = phi*phi2
+
+    return phif
+end
+
+
+function small_reduce(MRS::MatroidRealizationSpace) 
+    
+    !MRS.realizable && return MRS
+    phi = small_reduce_full_rec(gens(MRS.defining_ideal))
+    new_ring = codomain(phi)
+    new_ideal = phi(MRS.defining_ideal)
+    
+    new_ideal_gens = filter!(x->!iszero(x), phi.(gens(MRS.defining_ideal)))
+    
+    if length(new_ideal_gens) == 0
+        new_ideal = ideal(new_ring, [new_ring(0)])
+    else
+        new_ideal = ideal(new_ideal_gens)
+    end
+
+    new_mat = phi.(collect(MRS.realization_matrix))
+    new_mat = matrix(new_ring, new_mat)
+    new_ineq = unique!(filter(x->!is_unit(x), phi.(MRS.inequations)))
+    new_ineq = gens_2_factors(new_ineq)
+    
+    return MatroidRealizationSpace(new_ideal, new_ineq, new_ring, new_mat, MRS.F, MRS.char, MRS.q)
+    
+end
+
+=#
+
+#####################
+# full reduction    #
+#####################
+
+# computes the coefficient of v in monomial m. 
+function coefficient_monomial(v::RingElem, m::RingElem)
+    isone(degree(m,v)) || return "The variable is not a degree 1 factor."
+    mf = factor(m)
+    mfdict = Dict(mf)
+    u = unit(mf)
+    not_v = [k^(mfdict[k]) for k in keys(mfdict) if k != v ]
+    length(not_v) == 0 ? u : u*prod(not_v)
+end
+
+# computes the coefficient of v in f. 
+function coefficient_v(v::RingElem, f::RingElem)
+    isone(degree(f,v)) || return "degree of variable must be 1"
+    withv = [term(f,i) for i in 1:length(f) if v in vars(monomial(f,i))]
+    return sum([coefficient_monomial(v,m) for m in withv])
+end
+
+
+function find_solution_v(v::RingElem, Igens::Vector{<:RingElem}, 
+                         Sgens::Vector{<:RingElem}, R::MPolyRing) 
+
+    
+    with_v_deg_1 = [g for g in Igens if isone(degree(g,v))] 
+    length(with_v_deg_1) != 0 || return "can't isolate"
+
+    for f in with_v_deg_1
+
+        den = coefficient_v(v, f)
+        fac_den = poly_2_factors(den)
+        !issubset(fac_den, Sgens) && continue
+
+        no_v = [term(f,i) for i in 1:length(f) if !(v in vars(monomial(f,i)))]
+        iszero(length(no_v)) && continue
+              
+        h = R(-1)*sum(no_v)
+        return h//den
+        
+    end
+    return "can't solve for v"
+end
```

Comment:
What kind of return values are these? The signature is unclear. Why not throw errors if no solution can be found? Or return a pair `(solution_found::Bool, solution)` such that other functions can properly deal with this? You probably don't want someone to try `isa String` on the return value to figure out what is going on.

Reply:
changed so that an error is thrown. Now strings cannot be a return value.